amplitude/storages/inmemory_event_storage.go

```diff
@@ -1,34 +1,138 @@
 package storages
 
 import (
+	"sort"
+	"sync"
+	"time"
+
 	"github.com/amplitude/analytics-go/amplitude/types"
 )
 
-func NewInMemoryEventStorage(capacity int) types.EventStorage {
+func NewInMemoryEventStorage(chunkSize, sizeDivider int) types.EventStorage {
+	if sizeDivider < 1 {
+		sizeDivider = 1
+	}
+
+	currentChunkSize := chunkSize / sizeDivider
+	if currentChunkSize < 1 {
+		currentChunkSize = 1
+	}
+
 	return &inMemoryEventStorage{
-		capacity: capacity,
-		events:   make([]*types.Event, 0, capacity),
+		chunkSize:        chunkSize,
+		currentChunkSize: currentChunkSize,
+		sizeDivider:      sizeDivider,
+		events:           make([]*types.Event, 0, currentChunkSize),
 	}
 }
 
 type inMemoryEventStorage struct {
-	capacity int
-	events   []*types.Event
+	chunkSize        int
+	currentChunkSize int
+	sizeDivider      int
+	events           []*types.Event
+	retriedEvents    []*types.Event
+
+	mu sync.RWMutex
+}
+
+func (s *inMemoryEventStorage) PushNew(event *types.Event) {
+	s.push(false, event)
 }
 
-// Push pushes an event to the storage.
-func (i *inMemoryEventStorage) Push(event *types.Event) {
-	i.events = append(i.events, event)
+func (s *inMemoryEventStorage) ReturnBack(events ...*types.Event) {
+	s.push(true, events...)
 }
 
-// Pull returns all Events and empties EventStorage.
-func (i *inMemoryEventStorage) Pull() []*types.Event {
-	events := i.events
-	i.events = make([]*types.Event, 0, i.capacity)
+func (s *inMemoryEventStorage) push(prepend bool, events ...*types.Event) {
+	if len(events) == 0 {
+		return
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	prependIndex := 0
+	for _, event := range events {
+		if event.RetryAt.IsZero() {
+			s.addNonRetriedEvent(event, prepend, &prependIndex)
+		} else {
+			s.addRetriedEvent(event)
+		}
+	}
+}
+
+// PullChunk returns a chunk of events.
+func (s *inMemoryEventStorage) PullChunk() []*types.Event {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if len(s.events) >= s.currentChunkSize {
+		events := make([]*types.Event, s.currentChunkSize)
+		copy(events, s.events)
+		copy(s.events, s.events[s.currentChunkSize:])
+		s.events = s.events[:len(s.events)-s.currentChunkSize]
+
+		return events
+	}
+
+	events := make([]*types.Event, len(s.events), s.currentChunkSize)
+	copy(events, s.events)
+	s.events = s.events[:0]
+
+	now := time.Now()
+	retriedCount := 0
+	for retriedCount < len(s.retriedEvents) && len(s.events)+retriedCount < s.chunkSize && s.retriedEvents[retriedCount].RetryAt.Before(now) {
```

Comment:
I think any retry logic/ checks should also be maintained in the response processor.

Reply:
It is not retry logic/checks. I can rename `retry` to `pending` here